### PR TITLE
Upgrade KernelAbstractions to 0.8

### PR DESCRIPTION
**Pull request by vchuravy**

### CODE REVIEW

**Comment by vchuravy** [comment]
Ideally we would update to CUDA 4 at the same time but for that we need to update AMGX.

**Comment by navidcy** [comment]
https://github.com/JuliaGPU/AMGX.jl/issues/26

**Comment by navidcy** [comment]
@vchuravy I also an attempt to upgrade in #2782. If you feel that's useful then pick up that one?

**Comment by navidcy** [comment]
I didn't know about the `CUDAKernels.CUDADevice(;always_inline=true)` while working on #2782. Nice one!

**Comment by vchuravy** [comment]
@navidcy well I just invented that one. Not yet merged.

**Comment by vchuravy** [comment]
Maybe the best would be for you to merge this with your older PR?

**Comment by navidcy** [comment]
you mean open a PR from ncc/bump-kernel-abstractions-cudakernels branch to this one?

**Comment by navidcy** [comment]
I changed base on #2782... Is this what you meant @vchuravy?

**Comment by navidcy** [comment]
> you mean open a PR from ncc/bump-kernel-abstractions-cudakernels branch to this one?

ok, I guess you mean merge this _to_ #2782 

well if this PR works then we just close #2782 :)

**Comment by vchuravy** [comment]
Locally I am seeing:

```
Binary operations [GPU]: Error During Test at /home/vchuravy/src/Oceananigans/test/test_abstract_operations.jl:121
  Test threw exception
  Expression: ZeroField() - u == -u
  Scalar indexing is disallowed.
  Invocation of getindex resulted in scalar indexing of a GPU array.
  This is typically caused by calling an iterating implementation of a method.
  Such implementations *do not* execute on the GPU, but very slowly on the CPU,
  and therefore are only permitted from the REPL for prototyping purposes.
  If you did intend to index this array, annotate the caller with @allowscalar.
  Stacktrace:
    [1] error(s::String)
      @ Base ./error.jl:35
    [2] assertscalar(op::String)
      @ GPUArraysCore ~/.julia/packages/GPUArraysCore/B3xv7/src/GPUArraysCore.jl:100
    [3] getindex(::CuArray{Float64, 3, CUDA.Mem.DeviceBuffer}, ::Int64, ::Int64, ::Int64)
      @ GPUArrays ~/.julia/packages/GPUArrays/g2pOV/src/host/indexing.jl:9
    [4] getindex
```

**Comment by glwagner** [comment]
> Locally I am seeing:
> 
> ```
> Binary operations [GPU]: Error During Test at /home/vchuravy/src/Oceananigans/test/test_abstract_operations.jl:121
>   Test threw exception
>   Expression: ZeroField() - u == -u
>   Scalar indexing is disallowed.
>   Invocation of getindex resulted in scalar indexing of a GPU array.
>   This is typically caused by calling an iterating implementation of a method.
>   Such implementations *do not* execute on the GPU, but very slowly on the CPU,
>   and therefore are only permitted from the REPL for prototyping purposes.
>   If you did intend to index this array, annotate the caller with @allowscalar.
>   Stacktrace:
>     [1] error(s::String)
>       @ Base ./error.jl:35
>     [2] assertscalar(op::String)
>       @ GPUArraysCore ~/.julia/packages/GPUArraysCore/B3xv7/src/GPUArraysCore.jl:100
>     [3] getindex(::CuArray{Float64, 3, CUDA.Mem.DeviceBuffer}, ::Int64, ::Int64, ::Int64)
>       @ GPUArrays ~/.julia/packages/GPUArrays/g2pOV/src/host/indexing.jl:9
>     [4] getindex
> ```

Ok, I'll take a look

File: src/ImmersedBoundaries/immersed_boundary_condition.jl [hunk 1]
```diff
@@ -197,8 +197,8 @@ end
     q̃ᴮ = bottom_ib_flux(i, j, k, ibg, bc.bottom, loc, c, closure, K, id, clock, fields)
     q̃ᵀ =    top_ib_flux(i, j, k, ibg, bc.top,    loc, c, closure, K, id, clock, fields)
 
-    iᵂ, jˢ, kᴮ = index_right.((i, j, k), loc)
-    iᴱ, jᴺ, kᵀ = index_left.((i, j, k), loc)
+    iᵂ, jˢ, kᴮ = map(index_right, (i, j, k), loc) # Broadcast instead of map causes inference failure
```
Comment (glwagner) [inline src/ImmersedBoundaries/immersed_boundary_condition.jl:200]: 💡

Comment (vchuravy) [reply src/ImmersedBoundaries/immersed_boundary_condition.jl:200]: broadcast is a tricky to infer functions, map is much easier

**Review by glwagner — APPROVED** [review]

File: src/ImmersedBoundaries/immersed_boundary_condition.jl (hunk 1 above)
Comment (glwagner) [reply src/ImmersedBoundaries/immersed_boundary_condition.jl:200]: makes sense given the complexity of broadcast

**Comment by vchuravy** [comment]
LGTM and then I can work on CUDA v4

**Pull request merged**